### ISSUE TITLE
Track statistics for non-blocked requests.

### ISSUE THREAD

**Issue by yvrdarb** [CIDRAM/CIDRAM#204]
`abuseipdb➡lookup_everything:TRUE` does NOT force 100% of requests to be looked up, only a few as indicated from the API log.  Is there any way to force 100% requests get looked up?

Thanks for the great software.

**Comment by Maikuolan** [comment]
> Thanks for the great software.

You're welcome. :-)

> does NOT force 100% of requests to be looked up

Just to confirm: For the requests where lookups didn't occur, were the requests already blocked due to existing signatures, or did they manage to get through without being blocked?

**Comment by yvrdarb** [comment]
404s are still getting served when they are listed in abuse db and above my threshold setting.

So judging by the low number of checks done via the abuse db and high number of 404s, I have assumed that everything is not being filtered.

I am getting hit with a large number of POST /.env, GET /.env, wp-login.php,xmlrpc.php and wlwmanifest.xml requests that I would ideally like to filter.

**Comment by yvrdarb** [comment]
requests for robots.txt also get served (200) when they are listed in the abuse db and above my set threshold.

**Comment by yvrdarb** [comment]
Forget my comment, I later realized why; pretty simple ...

**Comment by Maikuolan** [comment]
No worries. :-)

> I later realized why

Do you mean just in regards to your comments prior to that (e.g., robots.txt, the .env files, etc), or in regards to the issue more generally?

**Comment by 737simpilot** [comment]
So it looks like you fixed it, though it would be nice to know what that was for the sake of others. 

I can tell you off the bat that CIDRAM only protects PHP files as far as I know. So if these 404s are images, XML files and text documents like the aforementioned robots.txt file, those won't be protected. 

For the POST stuff and what have you, you may want to read about, and understand how to fully utilize the auxiliary rules afforded to you. I have an auxiliary rule that states if the HTTP request is a POST (has to be in caps), and it's not equal to this list of directories, then block. I run a phpBB board and went through everything that requires a POST to create my list of not equals rules. Anything else that is summoned with a POST request is banned (with a captcha at the moment for any possible false positives). The other beauty here is that since my home IP is whitelisted in CIDRAM I'm bypassed so I didn't have to add white listed paths pertaining to the admin control panel seen as how I'm the only one that has access to that. So if someone were try to and weasel there way into the ACP with a POST request it'll block. On top of that I've always had the adm folder (where the ACP PHP files are at) protected with an htaccess file that only allows my IP. Since my IP doesn't change unless I change it myself (long story) it suffices. 

Re: the robots. 

Here you'll want to omit robots from indexing the CIDRAM folder or maybe the Ninjafirewall folder if you have it and other security scripts via htaccess. Now of course anybody can read the robots.txt file with a spoofed UA. So for me personally I only allow Google bot, Bing bot and Yahoo bot to access the robots.txt. Nothing else can see it. How this is accomplished is by two methods. 1) I have htaccess code that only allows those three robots based on their UA. And 2) If someone were to spoof the Google bot UA et al they'll get blocked by CIDRAM due to CIDRAM's search engine verification via rDNS. So _no soup for you_ if you're some residential IP, cloud or hoster IP using a spoofed Google bot UA or what ever.

This is what the htaccess looks like:

```
#Check that the request is for /robots.txt
RewriteCond %{REQUEST_URI} ^/robots.txt
#Check that the request matches an existing file
RewriteCond %{REQUEST_FILENAME} -f
# Check that the user agent does not contain google etc
RewriteCond %{HTTP_USER_AGENT} !google
RewriteCond %{HTTP_USER_AGENT} !yahoo
RewriteCond %{HTTP_USER_AGENT} !bing
# If all conditions above are met, then deny access to this request
RewriteRule ^ - [F,L]
```


You can of course tailor this to other bots at your will. The code is of course meant for Apache or Litespeed. You'd have to find an equivalent for Nginx.

The other thing here is I wouldn't use the default folder name CIDRAM in the public_html space. I renamed my CIDRAM and Ninjafirewall folders as I believe obscurity through security is a layer. Especially since I told the good bots not to index those folders and my _Google search dorking_ hasn't yielded those folders in a Google search.

Addendum. 

Those paths of xmlrpc.php, wlwmanifest.xml, etc are very common paths to be searched for looking for WordPress vulnerabilities. So if you're not running WordPress it may be in your best interest to block them in the auxiliary rules using wild cards.

**Comment by yvrdarb** [comment]
Thanks guys; lots of good information for me to digest 737.

Yes, it didn't quite click at the time; the fact that it only protects php scripts.

Basically I have about 800 domains that all resolve to one ip address, so lots of attention focused at one point, all resolving to a php lander script.  95% of all traffic is stopped at the firewall and I use CIDRAM as a final filtering tool.  Combined, both let me breeze through a DDOS attack without breaking a sweat a couple weeks back.  I would surmise that some providers pass the complainants originating IP address back to the person running the VPS when abuse complaints are made.

Also, a quick suggestion: it would be handy if it kept a running count of requests permitted and requests blocked.

**Comment by Maikuolan** [comment]
> Also, a quick suggestion: it would be handy if it kept a running count of requests permitted and requests blocked.

Could be a nice additional statistic to have available. I'll look further into it tomorrow. :-)

**Comment by 737simpilot** [comment]
So are you running this as part of your hosting business? I just can't even imagine one user owning 800 domains - and under one IP address! LOL 800 domains is at least $8,000 a year. 

Having said that, I don't think CIDRAM was created for such a use case but more for individual website owners and their own website's. So if you are using CIDRAM for a web hosting situation, it could have some potential drawbacks by virtue of severally limiting your customer's abilities and what not I think. 

I suppose if I had a web hosting business and used CIDRAM I'd use it only at its bare minimum. As it is, CIDRAM can and will block websites that test web page loading speed and other things unless you create specific whitlists. So your customers may not be able to use such a website or even the _Is it down right now_ website.

If you aren't a web hosting business, then that is a mighty impressive portfolio of domains!  I've heard of people using many domains to help increase SEO for one targeted website and what not, but not by the hundreds. LOL For me, I own about ~7 domains right now.  Couple of which I listed for sale since I think they _might_ have sales potential.

**Comment by yvrdarb** [comment]
Sorry for the delayed response, this is just a sideline hobby for me.

No, not a hosting business, just a lowly domainer.  It works well, perhaps ideally for my configuration; all http requests resolve to a glorified php contact form which is then filtered by CIDRAM.  Apart from my previously noted general concerns about 404 requests the only other issues is the occasional spam that comes in from my form.

I have often wondered why CIDRAM isn't much more popular; for shared hosting it is probably the best defence when combined with a WAF.  Perhaps better marketing is required ;)

**Comment by 737simpilot** [comment]
Believe me. I have advertised CIDRAM and Ninjafirewall like no tomorrow at the phpBB website adn maybe other places. I refereed to those two scrips in multiple topics and they were mentioned in my forum signature. Hardly anyone cared or were just too myopic to know how to use them. Especially judging by their level of intellect with their forum questions. It made me wonder why they deployed a website in the first place. 

Perhaps a few users used it. I know one deployed Ninjafirewall and I know another user sent me a PM/DM thanking me for mentioning CIDRAM. I do have a write up about Ninjafirewall at my website, but I don't have anything on CIDRAM. I should write something. 

I don't run SEO tools, use analytics and contribute to my website all the time for SEO, so I may need to try and fix that. 

Just curious. How did you become aware of CIDRAM?


Edit- 

Here's a list of everyone that's forked CIDRAM. IP-CAM is damn interesting. I've never seen so man repositories! LOL https://github.com/CIDRAM/CIDRAM/network/members

**Comment by Maikuolan** [comment]
Probably worth mentioning, I guess, that CIDRAM *might* be more popular than it overtly appears, though we'd have no real way to confirm it, one way or the other. Since many people often download, install, and use software, without ever posting anything on a support tracker, mentioning anything on a forum anywhere, telling anyone they use it, etc; or here at GitHub, cloning a repository and using the software or code associated with that repository, without bothering to give it a star, nor bothering to subscribe to and watch the repository; and since I don't implement any kind of tracking technologies, "web-bugs", or anything else into the codebase for that matter capable of giving us reliable figures, in short.. I have no real idea how popular or unpopular CIDRAM actually is. Still.. As long as it serves a purpose, and is useful for at least those of us that I'm aware of, eh.. good enough, I suppose. :-)

**Comment by 737simpilot** [comment]
Yeah, I totally forgot about the stars. There are 86 as of this post. https://github.com/CIDRAM/CIDRAM/stargazers

**Comment by yvrdarb** [comment]
Sorry, email got buried ...

> Just curious. How did you become aware of CIDRAM?

Can't recall, it has been a little while; 18 months perhaps since I have been using it.  I first came across it when I was on shared hosting while I was disparately searching for an effective defence.  Now I am on a VPS with the added tool of a firewall and the ability to install other security software.

I have thrown mention of CIDRAM out there a couple of times on forums but didn't seem to get any attention.

BTW, a quick question for an access log:

```
- 24.78.139.XXX - [08/Apr/2021:05:08:24 +0000] "GET / HTTP/1.1" 200 1620 - "Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1"

- 24.78.139.XXX - [08/Apr/2021:05:08:25 +0000] "GET /favicon.ico HTTP/1.1" 200 129122 - "MobileSafari/604.1 CFNetwork/1220.1 Darwin/20.3.0"
```

The UA change puts it at a 99.99% chance of being a bot?

A "normal" hit on my index.php results in three requests, such as below which would be another flag IMHO?

```
- 99.244.68.XXX - [08/Apr/2021:03:33:46 +0000] "GET / HTTP/1.1" 200 1620 - "Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1"

- 99.244.68.XXX - [08/Apr/2021:03:33:46 +0000] "GET /captcha.php?rand=1961591986 HTTP/1.1" 200 2828 http://www.victoriaXXXXX.ca/ "Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1"

- 99.244.68.XXX - [08/Apr/2021:03:33:47 +0000] "GET /favicon.ico HTTP/1.1" 200 129122 http://www.victoriaXXXXXX.ca/ "Mozilla/5.0 (iPhone; CPU iPhone OS 14_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1"
```

Basically I am getting lots of questionable (clean in abuse db) hits from our three largest internet providers here which I am trying to figure out and filter.  Perhaps I am just getting too obsessive though ...

**Comment by 737simpilot** [comment]
That looks like normal traffic. The `CFNetwork/1220.1` part is due to something in iPhones which I can't remember off hand. I think I read about it at Stack Exchange. I've seen it myself on my end.  However, do note that UAs can be forged. So just putting that out there. 

Since the IPs look like legit Canadian residential IPs, it's probably fine. However, residential IPs can be the source of shenanigans if that customer gets their router/modem hijacked and turned into a zombie botnet. Four ways to help rule this out:

1) Check the IP at Shodan. If there's an open port which is like 75xx something (can't remember), that port is typically a port that is open that the user doesn't know about and _can be_ hijacked by a hacker turning the router/modem into a zombie botnet. Many routers and modems have this port open so it's not proof positive this may be the case, but could aid in the detective work. 

2) You can also check the IP at Censys. Sometimes Censys shows open ports that Shodan doesn't and vice versa.  

3) Check the IP at Greynoise. If the residential IP is turned into a zombie botnet, it could be used for purposes that Greynoise will pick up on and list what that IP is doing. 

4) Check the IP at AbuseIPDB. Even one hit could help in the detective work. It may be so new that you're just a few that are seeing it. This has been my case a few times. 

In the case of Shodan, Censys and Greynoise, neither work with IPv6 and that's a real wrench in the gears when I see IPv6 addresses.  AbuseIPDB and some other websites will have to be used. Scamalytics is another one, but they're not as reliable as I'd like them to be. I've found many IPs listed like hell at AbuseIPDB to only have a 0 - 2 rating at Scamalytics. 

Now if you see a residential IP have open web ports like 80, 443, 22, and 21, etc, that could also aid in the detective work. But some people might be running their own server or a Plex server, etc. Though, Plex I think uses a different port. But I'm sure that port can be changed in the Plex server software. The banner from Shodan and Censys should say what the port is being used for. Googling the acronyms used in the banner can tell you more. Like NTP. Network Time Protocol. 

Sometimes I wonder if it'd be prudent for CIDRAM to do a web socket check on the user's connection for known ports as an informative use case only to investigate the possibility of shenanigans. Not sure if possible, ethical or what ever, but I know eBay is doing this looking for VNC ports and what not to help them combat hacked accounts. Since I'm very privacy/security focused, I blocked any and all web socket connections in my browser using uBlock Origin. If you use uBlock Origin, just add the following to the "My Filters List." 
`# Blocks websockets like eBay local port scan crap.`
`||ws://`
`||wss://`

This may or may not light up the radar so to speak for eBay et al since you're blocking web sockets. Because you might look suspicious and the risk of a captcha challenge would be greater.

**Comment by Maikuolan** [comment]
*"Track statistics for non-blocked requests."*

Earmarking for v3, since I have an idea in mind that'll be able to achieve this, but which is also "backwards-incompatible" (affects existing configuration directives).

**Comment by Maikuolan** [comment]
Implemented (9234a11b363f7ed7dfe4607ca63ad5c494a2a250). :-)

*Screenshot for the updated `statistics` configuration directive (v3):*
![statistics](https://user-images.githubusercontent.com/12571108/156397518-cca8edbf-1670-4a0c-a994-db29ec95285e.png)

Seeing as v3 is still an alpha (i.e., not yet stable enough to be safely used in production), I wouldn't recommend updating to it just yet, and it may still be a little while yet before we're able to actually *benefit* from the changes, but at least it's done now (so, as soon as it enters beta, or as soon as we see a `v3.0.0` tag appear here at the repository, go for it). Can confirm the relevant changes are working properly at my end via local testing.

*(Also, hooray for being the first v3-specific issue to be resolved)! 🎉* ^^

**Comment by 737simpilot** [comment]
> Yeah, I totally forgot about the stars. There are 86 as of this post. https://github.com/CIDRAM/CIDRAM/stargazers

93 now. Up by 7 in one year to the ~ month.
